lib/heuristics/concerns/periodic_end_phase.rb

```diff
@@ -106,52 +65,49 @@ def add_missing_visits
       vehicle_id = best_cost[1][:vehicle]
       log "It is interesting to add #{id} at day #{day} on #{vehicle_id}", level: :debug
 
-      @ids_to_renumber |= [id]
       insert_point_in_route(@candidate_routes[vehicle_id][day], best_cost[1])
-      @output_tool&.add_single_visit(day, @services_data[id][:used_days], id, @services_data[id][:raw].visits_number)
+      @output_tool&.add_single_visit(day, @services_assignment[id][:days], id, @services_data[id][:raw].visits_number)
 
       costs = update_costs(costs, best_cost)
     end
-
-    reaffect_visits_number
   end
 
-  def update_costs(costs, best_cost)
-    # update costs for inserted id, available_days changed
-    uninserted_set = @uninserted.select{ |_key, info| info[:original_id] == best_cost[0] }.keys
-    @uninserted.delete(uninserted_set.first)
-    if uninserted_set.size > 1
-      available_days = days_respecting_lapse(best_cost[0], @candidate_routes[best_cost[1][:vehicle]])
-
-      cost = find_best_day_cost(@candidate_routes[best_cost[1][:vehicle]], best_cost[0], available_days)
-
-      if cost
-        costs[best_cost[0]] = cost
-      else
-        costs.delete(best_cost[0])
-      end
+  # Updates costs after insertion of inserted_cost
```

Comment:
Do you mean `inserted_id` instead of `inserted_cost` ?
```suggestion
  # Updates costs after insertion of inserted_id
```